.travis.yml

```diff
@@ -1,6 +1,6 @@
 sudo: false
 language: cpp
-dist: trusty
```

Comment:
Why the change?

Ddid we change the minimum C++ standard version or something?

Reply:
FYI, I'm using mqtt_cpp on OpenWRT, which only has GCC 7.3, so I wouldn't be able to use the latest release if we need GCC 9 now.

Reply:
g++ outputs compile error due to g++'s bug. I need to avoid it.
I don't know how to install g++9 on trusty so I updated it.

Reply:
What was the compiler error? That's really weird, I did see any code that looked problematic?

Reply:
The compile error is happened on the following code:

```cpp
[
    self = shared_from_this()
]
```

https://travis-ci.org/redboltz/mqtt_cpp/jobs/572260184#L1050

```
/home/travis/build/redboltz/mqtt_cpp/include/mqtt/endpoint.hpp:9949:52: error: cannot call member function ‘std::shared_ptr<_Tp> std::enable_shared_from_this<_Tp>::shared_from_this() [with _Tp = mqtt::endpoint<std::mutex, std::lock_guard, 2>]’ without object
                             self = shared_from_this(),
                                    ~~~~~~~~~~~~~~~~^~
```

I can avoid it adding redundant `this->` as follows:

```cpp
[
    self = this->shared_from_this()
]
```

However, if I do that, then MSVC reports the following compile error:

https://redboltz.visualstudio.com/redboltz/_build/results?buildId=469

```
  C:\Program Files\Boost\1.69.0\include\boost-1_69\boost/pending/integer_log2.hpp(7): note: This header is deprecated. Use <boost/integer/integer_log2.hpp> instead.
  Please define _WIN32_WINNT or _WIN32_WINDOWS appropriately. For example:
  - add -D_WIN32_WINNT=0x0501 to the compiler command line; or
  - add _WIN32_WINNT=0x0501 to your project's Preprocessor Definitions.
  Assuming _WIN32_WINNT=0x0501 (i.e. Windows XP target).
D:\a\1\s\include\mqtt/endpoint.hpp(9982): error C2039: 'shared_from_this': is not a member of 'mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::process_connect_impl::<lambda_f447512e4e7450f0685e2f392f3b8c35>' [D:\a\1\s\build\test\as_buffer_async_pubsub_1.vcxproj]
          with
          [
              Socket=mqtt::tcp_endpoint<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>,boost::asio::io_context::strand>
          ]
  D:\a\1\s\include\mqtt/endpoint.hpp(9983): note: see declaration of 'mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::process_connect_impl::<lambda_f447512e4e7450f0685e2f392f3b8c35>'
          with
          [
              Socket=mqtt::tcp_endpoint<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>,boost::asio::io_context::strand>
          ]
  D:\a\1\s\include\mqtt/endpoint.hpp(9828): note: while compiling class template member function 'void mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::process_connect_impl(mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::async_handler_t,mqtt::buffer,mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::connect_phase,mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::connect_info &&)'
          with
          [
              Socket=mqtt::tcp_endpoint<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>,boost::asio::io_context::strand>
          ]
  D:\a\1\s\include\mqtt/endpoint.hpp(9813): note: see reference to function template instantiation 'void mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::process_connect_impl(mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::async_handler_t,mqtt::buffer,mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::connect_phase,mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>::connect_info &&)' being compiled
          with
          [
              Socket=mqtt::tcp_endpoint<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>,boost::asio::io_context::strand>
          ]
  D:\a\1\s\include\mqtt/client.hpp(40): note: see reference to class template instantiation 'mqtt::endpoint<Socket,std::mutex,std::lock_guard,2>' being compiled
          with
          [
              Socket=mqtt::tcp_endpoint<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>,boost::asio::io_context::strand>
          ]
```

It seems that MSVC misunderstood that `this` is internally created lambda class object.

To solve both incomplete compiler problems, I remove the redundant `this->` and update g++.

Reply:
I normally want to use the latest compiler version, but I'm currently limited to GCC 7.3.0 (I am, of course, always working on ways to upgrade to newer compilers..., but currently not able to), so I hope we can find a different work around.

Why not do:

`auto self = shared_from_this();`

before creating the lambda?

I also pointed out that there's a lot of places where we're calling shared_from_this() repeatedly, and could instead pass the shared_ptr into the function that then creates the lambda, avoiding the need for repeated calls to shared_from_this().

Reply:
Moved `shared_from_this()` call to outside of lambda expressions.
In addition, minimized `shared_from_this()` call.

Reply:
I downgrade g++ version to 7.4. Not only to test the workaround but also codecov support. I couldn't install gcov-9.